### PR TITLE
feat: generate types and docs for v: variables

### PR DESCRIPTION
**Pull request by lewis6991**
- doc: move vim-variables to separate file
- feat: generate types and docs for v variables

### CODE REVIEW

File: src/nvim/vvars.lua [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+    ]=],
+  },
+  option_command = {
+    type = 'string',
+    desc = [=[
+      Command used to set the option. Valid while executing an
+      |OptionSet| autocommand.
+        value        option was set via ~
+        "setlocal"   |:setlocal| or ":let l:xxx"
+        "setglobal"  |:setglobal| or ":let g:xxx"
```
Comment (clason) [inline src/nvim/vvars.lua:552]:
```suggestion
        "setlocal"   |:setlocal| or `:let l:xxx`
        "setglobal"  |:setglobal| or `:let g:xxx`
```

**Review by clason — APPROVED** [review]

File: runtime/lua/vim/_meta/vvars.lua [hunk 2]
```diff
@@ -1,23 +1,758 @@
...
+vim.v.completed_item = ...
+
 --- The count given for the last Normal mode command.  Can be used
 --- to get the count before a mapping.  Read-only.  Example:
+---
 --- ```vim
---- :map _x :<C-U>echo "the count is " .. v:count<CR>
---- ```
+---   :map _x :<C-U>echo "the count is " .. v:count<CR>
+--- <
```
Comment (lucario387) [inline runtime/lua/vim/_meta/vvars.lua:70]: This looks a bit confusing

File: src/nvim/vvars.lua [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+  },
+  count = {
+    type = 'integer',
+    desc = [=[
+      The count given for the last Normal mode command.  Can be used
+      to get the count before a mapping.  Read-only.  Example: >vim
+        :map _x :<C-U>echo "the count is " .. v:count<CR>
+      <
+      Note: The <C-U> is required to remove the line range that you
+      get when typing ':' after a count.
```
Comment (clason) [inline src/nvim/vvars.lua:81]:
```suggestion
      get when typing `:` after a count.
```
(or ":", but not single quotes)

**Comment by clason** [comment]
Note to self/others with too much time on their hands: `src/nvim/options.lua` could use some language annotations in the code blocks as well.

File: src/nvim/vvars.lua [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+      Normal mode command to be executed after a file has been
+      opened.  Can be used for a |SwapExists| autocommand to have
+      another Vim open the file and jump to the right place.  For
+      example, when jumping to a tag the value is ":tag tagname\r".
+      For ":edit +cmd file" the value is ":cmd\r".
+    ]=],
+  },
+  t_bool = {
+    type = 'integer',
+    tags = { 'v:t_TYPE' },
```
Comment (zeertzjq) [inline src/nvim/vvars.lua:741]: This should be moved to `t_blob` now.

File: src/nvim/vvars.lua [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+  errmsg = {
+    type = 'string',
+    desc = [=[
+      Last given error message.
+      Modifiable (can be set).
+      Example: >vim
+        let v:errmsg = ""
+        silent! next
+        if v:errmsg != ""
+          ... handle error
```
Comment (zeertzjq) [inline src/nvim/vvars.lua:147]:
```suggestion
          " ... handle error
```

File: src/nvim/vvars.lua (hunk 3 above)
Comment (zeertzjq) [reply src/nvim/vvars.lua:81]: Isn't single quotes fine for a single char?

File: scripts/gen_eval_files.lua [hunk 6]
```diff
@@ -534,6 +546,31 @@ local function render_option_meta(_f, opt, write)
   end
 end
 
+--- @param _f string
+--- @param opt vim.option_meta
+--- @param write fun(line: string)
+local function render_vvar_meta(_f, opt, write)
+  write('')
+
+  local desc = split(norm_text(opt.desc))
```
Comment (zeertzjq) [inline scripts/gen_eval_files.lua:555]: `norm_text()` should be called after removing leading spaces.

File: runtime/lua/vim/_meta/vvars.lua (hunk 2 above)
Comment (zeertzjq) [reply runtime/lua/vim/_meta/vvars.lua:70]: Because `norm_text()` should be called after removing leading spaces.

File: src/nvim/vvars.lua (hunk 3 above)
Comment (clason) [reply src/nvim/vvars.lua:81]: > Isn't single quotes fine for a **single** char?

Ah, you're right, I missed that (also for non-alphanumeric chars); I was worried that the vimdoc parser is not as lenient, but I just tested that it is.

(Have I mentioned how much I hate vimdoc and it's Pirate Code of a "specification"?)

File: src/nvim/vvars.lua [hunk 7]
```diff
@@ -0,0 +1,868 @@
...
+        row              Row count of popup menu on |CompleteChanged|,
+                         relative to screen.
+        col              Col count of popup menu on |CompleteChanged|,
+                         relative to screen.
+        size             Total number of completion items on
+                         |CompleteChanged|.
+        scrollbar        Is |v:true| if popup menu have scrollbar, or
+                         |v:false| if not.
+        changed_window   Is |v:true| if the event fired while
+        changing window  (or tab) on |DirChanged|.
```
Comment (zeertzjq) [inline src/nvim/vvars.lua:214]:
```suggestion
                         changing window  (or tab) on |DirChanged|.
```

File: runtime/doc/vvars.txt [hunk 8]
```diff
@@ -86,217 +94,247 @@ v:ctype		The current locale setting for characters of the runtime
...
+v:hlsearch
+		Variable that indicates whether search highlighting is on.
 		Setting it makes sense only if 'hlsearch' is enabled. Setting
 		this variable to zero acts like the |:nohlsearch| command,
-		setting it to one acts like >
-			let &hlsearch = &hlsearch
-<		Note that the value is restored when returning from a
+		setting it to one acts like
+		let &hlsearch = &hlsearch
+		Note that the value is restored when returning from a
```
Comment (zeertzjq) [inline runtime/doc/vvars.txt:318]: Missing codeblock markers

File: runtime/doc/vvars.txt [hunk 9]
```diff
@@ -385,69 +435,87 @@ v:oldfiles	List of file names that is loaded from the |shada| file on
...
+			*v:option_command* *option_command-variable*
+v:option_command
+		Command used to set the option. Valid while executing an
+		|OptionSet| autocommand.
+		  value        option was set via ~
+		  "setlocal"   |:setlocal| or ":let l:xxx"
+		  "setglobal"  |:setglobal| or ":let g:xxx"
+		  "set"        |:set| or |:let|
+		  "modeline"   |modeline|
+		      --
```
Comment (zeertzjq) [inline runtime/doc/vvars.txt]: ?

File: src/nvim/vvars.lua [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+    desc = [=[
+      Last given error message.
+      Modifiable (can be set).
+      Example: >vim
+        let v:errmsg = ""
+        silent! next
+        if v:errmsg != ""
+          " ... handle error
+      <
+
```
Comment (zeertzjq) [inline src/nvim/vvars.lua:149]:
```suggestion
```

File: runtime/doc/vvars.txt [hunk 11]
```diff
@@ -457,111 +525,131 @@ v:register	The name of the register in effect for the current normal mode
...
+		Primary listen-address of Nvim, the first item returned by
 		|serverlist()|. Usually this is the named pipe created by Nvim
 		at |startup| or given by |--listen| (or the deprecated
 		|$NVIM_LISTEN_ADDRESS| env var).
 
 		See also |serverstart()| |serverstop()|.
 		Read-only.
 
-								       *$NVIM*
+		                   *$NVIM*
```
Comment (zeertzjq) [inline runtime/doc/vvars.txt:560]: This tag's position is a bit strange.

File: runtime/doc/vvars.txt [hunk 12]
```diff
@@ -457,111 +525,131 @@ v:register	The name of the register in effect for the current normal mode
...
 		Unlike stdin and stdout (see |stdioopen()|), stderr is always
-		open for writing. Example: >
-			:call chansend(v:stderr, "error: toaster empty\n")
-<
-					*v:swapname* *swapname-variable*
-v:swapname	Name of the swapfile found.
-		Only valid during |SwapExists| event.
-		Read-only.
+		open for writing. Example:
+		:call chansend(v:stderr, "error: toaster empty\n")
```
Comment (zeertzjq) [inline runtime/doc/vvars.txt:596]: Missing code block markers

File: src/nvim/vvars.lua [hunk 13]
```diff
@@ -0,0 +1,869 @@
...
+    desc = [=[
+      The point where the exception most recently caught and not
+      finished was thrown.  Not set when commands are typed.  See
+      also |v:exception| and |throw-variables|.
+      Example: >vim
+       try
+         throw "oops"
+       catch /.*/
+         echo "Exception from" v:throwpoint
+       endtry
```
Comment (zeertzjq) [inline src/nvim/vvars.lua:802]:
```suggestion
        try
          throw "oops"
        catch /.*/
          echo "Exception from" v:throwpoint
        endtry
```

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**